### PR TITLE
An orchestration-compatible env_utils (C4-689. C4-815, C4-860)

### PR DESCRIPTION
**Pull request by netsettler**
This PR is responsive to [Need a configurable env_utils for orchestrated applications in isolated accounts (C4-689)](https://hms-dbmi.atlassian.net/browse/C4-689).

Now that [Rewrite uses of variables imported from dcicutils.env_utils (C4-700)](https://hms-dbmi.atlassian.net/browse/C4-700) (duplicated by [C4-690](https://hms-dbmi.atlassian.net/browse/C4-690) )is done, we hope this will be a compatible transition. It still is a major change, so I recommend ultimately bumping the major version number. (Technically it would be an incompatible change because the variables in question are merely deprecated, but I did a complete search of all our repos for uses, and made updates. There may yet be an issue in 4dn-cloud-infra because it's a private repo and might not have been picked up in the search.) Regardless, heavy testing is advised.

This does several things:

* Renames dcicutils/env_utils.py to dcicutils/env_utils_legacy.py so that old functionality is not disturbed.

* Introduces new file dcicutils/env_utils.py that wraps the legacy support, introducing a new configurable set of functions.

* Adds various kinds of tests.

### CODE REVIEW

File: test/test_es_utils.py [hunk 1]
```diff
@@ -104,10 +105,12 @@ def es_client_fixture(integrated_ff):
     return create_es_client(integrated_ff['es_url'])
 
 
+@pytest.mark.skip(reason="Direct ES search cannot work outside the firewall.")
 @pytest.mark.integrated
+@pytest.mark.direct_es_query
 def test_lucene_query_basic(es_client_fixture):
     """ Tests basic lucene queries via the underlying endpoint on mastertest """
-    results = execute_lucene_query_on_es(client=es_client_fixture, index='fourfront-mastertestuser', query={})
+    results = execute_lucene_query_on_es(client=es_client_fixture, index=f'{INTEGRATED_ENV_INDEX}user', query={})
```
Comment (willronchetti) [inline test/test_es_utils.py:113]: I see the use now - why do it this way though? You can get this info from the health page. The variable name is also not good... this is an ES index namespace.

File: test/test_ff_utils.py [hunk 2]
```diff
@@ -314,10 +321,10 @@ def test_unified_authentication_prod_envs_integrated_only():
...
     staging_auth = ff_utils.unified_authentication(ff_env="staging")
     assert staging_auth == ff_prod_auth
-    green_auth = ff_utils.unified_authentication(ff_env="fourfront-green")
-    assert green_auth == ff_prod_auth
-    blue_auth = ff_utils.unified_authentication(ff_env="fourfront-blue")
-    assert blue_auth == ff_prod_auth
+#     green_auth = ff_utils.unified_authentication(ff_env="fourfront-production-green")
+#     assert green_auth == ff_prod_auth
+#     blue_auth = ff_utils.unified_authentication(ff_env="fourfront-production-blue")
+#     assert blue_auth == ff_prod_auth
```
Comment (willronchetti) [inline test/test_ff_utils.py:327]: FYI, this should work with main account creds. They should also line up directly with `data` and `staging`.

File: test/test_ff_utils.py [hunk 3]
```diff
@@ -435,7 +442,7 @@ def test_stuff_in_queues_integrated(integrated_ff):
     for item in search_res[:8]:
         ff_utils.patch_metadata({}, obj_id=item['uuid'], key=integrated_ff['ff_key'])
     time.sleep(3)  # let queues catch up
-    stuff_in_queue = ff_utils.stuff_in_queues(integrated_ff['ff_env'], check_secondary=True)
+    stuff_in_queue = ff_utils.stuff_in_queues(integrated_ff['ff_env_index'], check_secondary=True)
```
Comment (willronchetti) [inline test/test_ff_utils.py:445]: ``ff_env_index_namespace`` is the name you want.

File: test/test_s3_utils.py [hunk 4]
```diff
@@ -92,28 +98,31 @@ def test_regression_s3_utils_short_name_c4_706():
...
-    env_url = beanstalk_utils.get_beanstalk_real_url(env)
-    health_page_url = f"{env_url}/health?format=json"
-    return requests.get(health_page_url).status_code == 200
+def _env_is_probably_up_and_healthy(env):
+    ignored(env)
+    return True  # not quite right, but
+    # This code won't work in container systems...
+    # env_url = beanstalk_utils.get_beanstalk_real_url(env)
+    # health_page_url = f"{env_url.rstrip('/')}/health?format=json"
+    # return requests.get(health_page_url).status_code == 200
```
Comment (willronchetti) [inline test/test_s3_utils.py:107]: Not clear on why you can't do this in the new paradigm. GLOBAL_ENV_BUCKET should be set, and thus we should be able to resolve the portal URL. The system does not function without GLOBAL_ENV_BUCKET, and we should not write code that tries to handle things when it is absent. Rather we should detect it is absent and report an error.

File: test/test_s3_utils.py [hunk 5]
```diff
@@ -529,13 +579,24 @@ def test_read_s3_zip_integrated(integrated_s3_info):
...
+    ffenv = integrated_names['ffenv'] if integrated_names else None
+
+    with mocked_s3utils_with_sse(beanstalks=[ffenv], require_sse=False):
+        s3_connection = s3Utils(env=ffenv)
+
+        # In our mock, this won't exist already on S3 like in the integrated version of this test,
+        # so we have to pre-load to our mock S3 manually. -kmp 13-Jan-2021
+        s3_connection.s3.upload_file(Filename=integrated_names['zip_path'],
+                                     Bucket=s3_connection.outfile_bucket,
+                                     Key=integrated_names['zip_filename'])
```
Comment (willronchetti) [inline test/test_s3_utils.py:593]: As stated previously, `s3Utils` does not implement a true interface to `s3`. This code while mocked appropriately would not function in production as it does not pass KMS arguments. Similar issue in other places here.

File: test/test_s3_utils.py [hunk 6]
```diff
@@ -908,12 +1006,26 @@ class MyS3(MockBotoS3Client):
...
+    # Now that we're containerized, there is always something in GLOBAL_ENV_BUCKET
+    #
+    # # These tests expect to be run in an environment that does not have these buckets bound globally.
+    # assert os.environ.get('GLOBAL_ENV_BUCKET') is None
+    # assert os.environ.get('GLOBAL_BUCKET_ENV') is None
+    #
+    # This is the rewrite:
+
+    global_env_bucket_for_testing = os.environ.get('GLOBAL_ENV_BUCKET')
+    assert global_env_bucket_for_testing  # probably it's 'foursight-envs', but let's be generic
```
Comment (willronchetti) [inline test/test_s3_utils.py:1018]: It needs to be `foursight-envs` IMO. If it's not you're trying to run utils tests in the wrong account and will get many failures. This is the type of thing that should be tested before kicking any other tests.

**Review by willronchetti — COMMENTED** [review]
Not ready for merge in current form - requires at mininum updates to include foursight result `bucket`

File: test/test_ff_utils.py (hunk 3 above)
Comment (netsettler) [reply test/test_ff_utils.py:445]: That would be OK.

File: CHANGELOG.rst [hunk 7]
```diff
@@ -7,6 +7,39 @@ Change Log
 ----------
 
 
+4.0.0
+=====
+
+* Configurable support for orchestrated CGAP in ``env_utils`` (C4-689).
```
[the remark this answers is not in the document]
Comment (netsettler) [reply CHANGELOG.rst]: I think at the time I wrote this entry you had not  yet expanded the scope of the work! Yeah, I'll fix that.

File: CHANGELOG.rst [hunk 8]
```diff
@@ -7,6 +7,39 @@ Change Log
...
+``"stage_mirroring_enabled"``    Should be ``"true"`` if mirroring is desired, otherwise omitted.
+``"orchestrated_app"``           This allows us to tell 'cgap' from 'fourfront', in case there ever is one.
+``"prd_env_name"``               The name of the prod env
+``"public_url_table"``           Dictionary mapping envnames & pseudo_envnames to public urls
+``"stg_env_name"``               The name of the stage env (or None)
+``"test_envs"``                  A list of environments that are for testing
+``"webprod_pseudo_env"``         The pseudo-env that is a token name to use in place of the prd env for shared
+                                 stg/prd situations, replacing ``fourfront-webprod`` in the legacy system.
+                                 (In orchestrations, this should usually be the same as the ``prd_env_name``.
+                                 It may or may not need to be different if we orchestrate the legacy system.)
```
[the remark this answers is not in the document]
Comment (netsettler) [reply CHANGELOG.rst]: Which bucket are we talking about?

File: dcicutils/__init__.py [hunk 9]
```diff
@@ -1,4 +1,8 @@
...
-from .ff_utils import *  # NOQA
-from .s3_utils import *  # NOQA
+# With this release, you need to import from beanstalk_utils, ff_utils, or s3_utils as appropriate. Don't rely on this.
+# In 4dn-dcic repos, there was only one use in dcicutils.test_base that has been updated.
+# In dbmi-bgm there were no uses.
+#
+# from __future__ import absolute_import
+# from .beanstalk_utils import *  # NOQA
+# from .ff_utils import *  # NOQA
+# from .s3_utils import *  # NOQA
```
[the remark this answers is not in the document]
Comment (netsettler) [reply dcicutils/__init__.py]: Yes, but we did this search the other day on a whole-github basis and made a list of what needs to change. That's why it's a major version bump though.

File: dcicutils/beanstalk_utils.py [hunk 10]
```diff
@@ -248,6 +249,11 @@ def _create_foursight_new(dest_env):
     fs['fs_url'] = get_foursight_env(dest_env, fs['bs_url'])
     fs['es_url'] = ff_utils.get_health_page(ff_env=dest_env)['elasticsearch']
     fs['foursight'] = create_foursight(**fs)
+    if is_orchestrated():
+        # TODO: Probably want to inherit some values from the old file in this case, since not all of those change.
+        raise NotImplementedError("Need to add orchestration support here.")
+    else:
+        fs['is_legacy'] = True
```
[the remark this answers is not in the document]
Comment (netsettler) [reply dcicutils/beanstalk_utils.py]: OK

File: CHANGELOG.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (netsettler) [reply CHANGELOG.rst]: Using "orchestrated C4 (CGAP and Fourfront)", as that's a little clearer to outsiders.
While there could be other such applications, I think that's low risk.

File: CHANGELOG.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (netsettler) [reply CHANGELOG.rst]: I'm putting in a `"foursight_bucket_table"` facility.

File: dcicutils/beanstalk_utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (netsettler) [reply dcicutils/beanstalk_utils.py]: The function `beanstalk_utils.swap_cname` has been removed.

File: test/conftest.py [hunk 11]
```diff
@@ -1,81 +1,120 @@
...
-                return health_json
-    except Exception as e:
-        raise RuntimeError("Unable to discover elasticsearch info for %s:\n%s: %s" % (envname, e.__class__.__name__, e))
+def _get_portal_url_for_testing(envname):
+    # This doesn't quite work, probably because fourfront_mastertest is not properly declared yet.
+    # portal_url = get_beanstalk_real_url(envname)
+    short_envname = remove_prefix('fourfront_', remove_prefix('fourfront-', envname))
+    print(f"short_envname={short_envname}")
+    portal_url = f"https://{short_envname}.4dnucleome.org"
+    return portal_url
```
[the remark this answers is not in the document]
Comment (netsettler) [reply test/conftest.py:23]: I've simplified a lot of this setup.

File: test/test_ff_utils.py (hunk 3 above)
Comment (netsettler) [reply test/test_ff_utils.py:445]: Done.

File: test/test_ff_utils.py (hunk 2 above)
Comment (netsettler) [reply test/test_ff_utils.py:327]: I've marked this test as needing repair in the future.

File: test/test_s3_utils.py (hunk 4 above)
Comment (netsettler) [reply test/test_s3_utils.py:107]: I've revised the way this works, but it's doing the proper test and doing the right thing. In fact, it was going to the wrong environment. It's better now.

**Comment by netsettler** [comment]
The version current right now is testing cleanly on my machine and also under GA with a manual invocation.

File: dcicutils/__init__.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (netsettler) [reply dcicutils/__init__.py:8]: I went through the layered PR and put all the breaking changes I saw into the CHANGELOG.

File: dcicutils/deployment_utils.py [hunk 12]
```diff
@@ -708,7 +708,7 @@ def build_ini_stream_from_template(cls, template_file_name, init_file_stream, *,
         }
 
         # if we specify an indexer name for bs_env, we did the deployment wrong and should bail
-        if bs_env in INDEXER_ENVS:
+        if is_indexer_env(bs_env):
```
Comment (netsettler) [inline dcicutils/deployment_utils.py:711]: @willronchetti, the function `is_indexer_env` will presently err so this code will fail. Would you rather I remove this clause or make `is_indexer_env` just unconditionally return `False`?`

File: dcicutils/misc_utils.py [hunk 13]
```diff
@@ -1506,6 +1506,25 @@ def do_subst(e):
     return exp
 
 
+class SingletonManager():  # TODO: Move to dcicutils
```
[the remark this answers is not in the document]
Comment (netsettler) [reply dcicutils/misc_utils.py:1509]: Done.

File: .github/workflows/main.yml [hunk 14]
```diff
@@ -45,4 +45,12 @@ jobs:
...
           poetry install
           poetry run flake8 dcicutils
-          make test
+          poetry run flake8 test
+          make test-for-ga
+
+      - name: Coveralls
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        if: ${{ matrix.python_version == '3.6' }}
```
[the remark this answers is not in the document]
Comment (netsettler) [reply .github/workflows/main.yml:54]: I'm leaving the support in but changing this conditional to run in 3.9 to see how that goes.

File: test/test_s3_utils.py [hunk 15]
```diff
@@ -123,7 +144,7 @@ def test_stg(ff_staging_envname):
             'url': util.url,
         }
         assert actual_props == {
-            'sys_bucket': 'elasticbeanstalk-fourfront-webprod-system',
+            'sys_bucket': 'elasticbeanstalk-fourfront-webprod2-system',
```
[the remark this answers is not in the document]
Comment (netsettler) [reply test/test_s3_utils.py:147]: OK. Done.

**Review by willronchetti — APPROVED** [review]
Giving final approval for merge and tag as 4.0.0

**Comment by netsettler** [comment]
The CHANGELOG has been adjusted to note that it is interim. In a later PR, I'll adjust the changelog for 4.0 to be more comprehensive. But this should be ready to go.

**Pull request merged**